File: src/cli/one-shot.ts
/**
 * One-shot CLI for running a single prompt against a DeepAgent.
 *
 * Used by `ask "question"` and `run <file>` subcommands.
 * No interactive loop — just send the prompt and print the response.
 */

import { readFileSync, existsSync } from "node:fs";
import { resolve } from "node:path";
import { createAppAgent } from "../runtime/agent-factory.js";
import { loadConfig } from "../runtime/config-loader.js";
import { logger } from "../runtime/logger.js";

const log = logger.child("one-shot");

export interface OneShotOptions {
  /** Path to config file */
  configPath?: string;
  /** Path to a custom system prompt file */
  promptPath?: string;
  /** Pre-set system prompt text */
  systemPrompt?: string;
  /** Workspace root directory */
  workspaceRoot?: string;
}

/**
 * Run a single prompt against an agent and print the response.
 */
export async function runOneShot(
  prompt: string,
  options: OneShotOptions = {}
): Promise<void> {
  const config = loadConfig({ configPath: options.configPath });
  const workspaceRoot = options.workspaceRoot || process.cwd();
  const systemPrompt = resolveSystemPrompt(options);

  log.info("Creating agent for one-shot prompt");
  const { agent } = createAppAgent(config, { cwd: workspaceRoot });

  const messages = [
    { role: "user" as const, content: systemPrompt },
    { role: "user" as const, content: prompt },
  ];

  try {
    const response = await agent.invoke({ messages });
    const content = extractContent(response);
    console.log(content);
  } catch (err) {
    console.error(`Error: ${err instanceof Error ? err.message : String(err)}`);
    process.exit(1);
  }
}

/**
 * Run a prompt from a file.
 */
export async function runPromptFile(
  filePath: string,
  options: OneShotOptions = {}
): Promise<void> {
  const fullPath = resolve(process.cwd(), filePath);
  if (!existsSync(fullPath)) {
    console.error(`Error: prompt file not found: ${fullPath}`);
    process.exit(1);
  }
  const prompt = readFileSync(fullPath, "utf-8").trim();
  await runOneShot(prompt, options);
}

// ─── Helpers (shared with REPL) ─────────────────────────

function resolveSystemPrompt(options: OneShotOptions): string {
  if (options.systemPrompt) {
    return options.systemPrompt;
  }

  if (options.promptPath) {
    const fullPath = resolve(process.cwd(), options.promptPath);
    if (existsSync(fullPath)) {
      return readFileSync(fullPath, "utf-8").replace(/^# .*\r?\n/, "").trim();
    }
  }

  const defaultPath = resolve(process.cwd(), "prompts/developer-agent.system.md");
  if (existsSync(defaultPath)) {
    return readFileSync(defaultPath, "utf-8").replace(/^# .*\r?\n/, "").trim();
  }

  return "You are a helpful DeepAgent assistant. Be concise and action-oriented.";
}

function extractContent(response: unknown): string {
  if (!response) return "(no response)";
  if (typeof response === "string") return response;

  if (Array.isArray(response)) {
    return response
      .map((m) => extractContent(m))
      .filter(Boolean)
      .join("\n");
  }

  const r = response as { messages?: unknown[]; content?: unknown; text?: unknown };
  if (Array.isArray(r.messages)) {
    return r.messages
      .map((m) => extractContent(m))
      .filter(Boolean)
      .join("\n");
  }

  if (typeof r.content === "string") return r.content;
  if (typeof r.text === "string") return r.text;

  if (Array.isArray(r.content)) {
    return r.content
      .map((b) => {
        if (typeof b === "string") return b;
        const block = b as { type?: string; text?: string };
        if (block.type === "text" && typeof block.text === "string") return block.text;
        return "";
      })
      .filter(Boolean)
      .join("\n");
  }

  try {
    return JSON.stringify(response, null, 2);
  } catch {
    return String(response);
  }
}

File: packages/deepagents-app-py/src/deepagents_app_py/surfaces/cli/one_shot.py
"""One-shot CLI — single prompt execution against the deepagents graph."""

from __future__ import annotations

import os
import sys
from pathlib import Path
from typing import Any

from deepagents_app_py.runtime.logger import logger


def _message_text(content: Any) -> str:
    """Extract plain text from a LangChain message ``content`` (str or blocks)."""
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts = []
        for block in content:
            if isinstance(block, dict) and block.get("type") == "text":
                parts.append(block.get("text", ""))
            elif isinstance(block, str):
                parts.append(block)
        return "".join(parts)
    return str(content)


def run_one_shot(prompt: str, options: Any = None) -> int:
    """Run a single prompt through the agent and print the final response."""
    log = logger.child("one-shot")
    log.info("Running one-shot prompt", prompt=prompt[:100])

    from deepagents_app_py.app.tools import collect_tools
    from deepagents_app_py.runtime.agent_config import build_graph
    from deepagents_app_py.runtime.config.config_loader import loadConfig

    ws = getattr(options, "workspace_root", None) or os.getcwd()
    config = loadConfig(
        {"workspaceRoot": ws, "configPath": getattr(options, "config_path", None)}
    )

    # One-shot: no checkpointer (no thread to persist).
    graph = build_graph(config, None, ws, collect_tools(), checkpointer=False)

    try:
        result = graph.invoke({"messages": [{"role": "user", "content": prompt}]})
    except Exception as exc:  # noqa: BLE001 — report cleanly to the user
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    messages = result.get("messages", []) if isinstance(result, dict) else []
    if messages:
        print(_message_text(messages[-1].content))
    return 0


def run_prompt_file(file_path: str, options: Any = None) -> int:
    """Read a prompt from a file and execute it."""
    log = logger.child("one-shot")
    try:
        content = Path(file_path).read_text(encoding="utf-8")
        log.info("Running prompt from file", file=file_path, length=len(content))
        return run_one_shot(content, options)
    except FileNotFoundError:
        print(f"Error: File not found: {file_path}", file=sys.stderr)
        return 1
    except OSError as exc:
        print(f"Error reading file {file_path}: {exc}", file=sys.stderr)
        return 1
